Libraries/JavaScriptAppEngine/System/JSTimers/JSTimersExecution.js

```diff
@@ -108,6 +117,34 @@ var JSTimersExecution = {
     }
   },
 
+  callIdleCallbacks: function(frameTime) {
+    // Arbitrary threshold: 10ms remaining in frame.
+    if (17 - (performanceNow() - frameTime) < 10) {
```

Comment:
If you assume that your normal workload is before this then you certainly will be close to the 16ms in which case eating into a frame would be dangerous. However that's just an artifact of so many events being processed before the animation is flushed. Most of that event handling such as handling network responses should actually be considered idle work.

One thing that made me shift my mental model was to think of these as requestNormalCallback instead of Idle. Because it is really not suppose to be just background work but most work.